Add error handling when host can not be reached

Ruben-Hartog opened the pull request.

Ruben-Hartog committed:
Add error handling to fetch transport

Ruben-Hartog committed:
Add onError to xhr requests

jonnyreeves commented:
Hey @Ruben-Hartog, thanks for opening this PR.  The integration tests are failing, would you mind taking a look please?

Ruben-Hartog commented:
Hi @jonnyreeves.

I am not sure why they are failing. I've tested it locally in chrome, firefox and safari and all tests pass. But somehow they are still failing when running on Travis. Do you have an idea?

jonnyreeves commented:
Hey @Ruben-Hartog - my guess would be how BrowserStack is handling the request to proxy to 'unkown-host.com'.  The code looks good, would just be nice if we can figure out how to test it 💃 

In the mean time, please can you update the Project's README to include information about the `onError()` callback?  Thanks!

Ruben-Hartog committed:
Update README.md

Ruben-Hartog committed:
Use a 'real' unknown address, address was actually known

Ruben-Hartog commented:
Hi @jonnyreeves ,

I have tried some different hosts, without success... They are successful on my local machine, so it could be something with BrowserStack. Are you able to run the tests on your local machine? Just to be sure.
Anyway, I have updated the README.

jonnyreeves commented:
Hey @Ruben-Hartog, as a result of #9 getting merged your branch now has conflict - if you could patch those up I will take a look at the tests locally to see if I can figure out a BrowserStack workflow.

If I can't work out a workflow then we can just disable that on the BrowserStack run (leaving it in place for local testing, which isn't ideal - but this is a valuable feature).

jonnyreeves commented:
... infact having written that - your original requirement may have been addressed with @MarcusLongmuir's switch to use the `loadend` event which catches both success and failure cases (maybe you just need to patch up the `fetchImpl`?

Ruben-Hartog commented:
Hi @jonnyreeves,

I resolved the conflicts.
Although the loadend event also catches errors, according to the documentation it is not possible to distinguish whether the transfer succeeded or not. And that's actually what we need 😉 .

jonnyreeves commented:
Hey @Ruben-Hartog - thanks for persevering - we'll get there!
1. Instead of adding a new method to the public API (`onError`) - do you think we could route the failure through to `onComplete` providing an `statusCode` of `0`?  This will make it much easier for consumers as they should always be checking the `statusCode` at the end of a request anyway.
2. It looks like we are not going to get a working integration tests for this feature - I don't suppose you fancy taking a stab at a unit test - if you're up for doing this, but don't feel comfortable bootstrapping a unit test suite let me know and I can lend a hand.

Cheers
Jonny.

Ruben-Hartog committed:
Use onComplete callback with status code 0 instead of onError

Ruben-Hartog commented:
Hi @jonnyreeves ,

That makes sense! I have replaced the onError with the onComplete callback which returns a statusCode of 0. I also removed the failing integration test.
If you like unit tests for this I suggest you to set up a unit test framework to your liking 😉 .

jonnyreeves commented on src/impl/mozXhr.js:
```diff
@@ -34,5 +34,10 @@ export default function mozXhrRequest(options) {
   }
   xhr.addEventListener('progress', onProgressEvent);
   xhr.addEventListener('loadend', onLoadEvent);
+  xhr.addEventListener('error', err => options.onComplete({
```
please can you pull this out into a helper methods (like `onProgressEvent` and `onLoadEvent`) for consistency.

jonnyreeves replied:
Pretty sure this should delegate through to `onRawComplete`?  @MarcusLongmuir would you mind taking a look please?

Ruben-Hartog committed:
Extract error handlers to functions

jonnyreeves commented:
Nice one - sorry for the back and forth!
